### DISCUSSION TITLE
Prisma Migration Example

### DISCUSSION THREAD

**Discussion by mcmar** [hasura/graphql-engine#7113] [General]
My company ([Vibely](https://www.crunchbase.com/organization/vibely)) is looking to migrate away from Prisma because they're abandoning their current customers with Prisma2.

We're facing a lot of challenges migrating to Hasura because the API is so different and we'd love to see any examples or documentation for how to migrate to Hasura. In particular, it's hard for us to figure out how to migrate mutations because the schema wraps everything in a `returning {` block, which means that we can't use `graphql-bindings` with our current schema because the `info` object won't translate. But maybe there's some tooling that I'm not aware of to translate that info object and wrap it with a `returning {` block?

It would be really helpful if there were a migration guide on how to do this. We're desperate to leave Prisma, but we're just not sure how to ease the migration path to Hasura 🙏

**Comment by mcmar** [comment]
I'm trying to migrate from Prisma with `prisma-binding` to Hasura with `graphql-binding`. I can setup all the tooling, but what I can't figure out is how to use the `info` object that's expecting a schema like `{ id, name }` with Hasura, which expects a schema like `{ returning { id, name } }`. I've never found a good tool to modify the info object, which means that I would have to modify my API's schema to also wrap everything inside `returning`, which would mean changing my frontend. I'd like to avoid changing the frontend, which means keeping my API schema the same, which means somehow wrapping/unwrapping the `info` object of my resolver with `returning`. If that's not possible, then there's no way for anybody to migrate smoothly from Prisma to Hasura.

**Comment by bkstorm** [comment]
I'm also migrating from Prisma 1 to Hasura. Prisma is not stable, they change a lot, and lack some important features. I have never use `graphql-binding`, and rarely use `info` object, currently I don't have any problem when migrating. Is it possible to stop using `graphql-binding` and find another tool?

**Comment by mcmar** [comment]
The issue isn't `graphql-binding` and I'm capable of setting up `graphql-binding` to use Hasura. It's an open-source tool that will work with any remote graphql schema.
The issue is that apollo-server uses resolvers. Those resolvers are passed an `info` object which contains information about the query including all the fields being queried. I currently pass that `info` object directly to `graphql-binding` because I've carefully designed my API's schema to be compatible with the prisma schema by following the OpenCRUD standard. Because Hasura doesn't support the OpenCRUD standard, that makes migration almost impossible without modifying the API schema, which then requires us to update all the clients that use that API schema and deprecate all the old clients. That's a total nonstarter. I'm hoping that instead of modifying the API schema, I can change the resolvers to modify the `info` object to match the HasuraCRUD schema, which means doing things like wrapping and unwrapping the `returning` object.

I found something promising with https://www.apollographql.com/docs/graphql-tools/schema-transforms/ but I'm not really sure how to go about implementing it and I'm not super familiar with the HasuraCRUD schema. It would be really cool to see a guide that uses Apollo's schema transforms to translate an OpenCRUD schema into a HasuraCRUD schema. I'm very motivated and willing to put in the effort to experiment and implement something like that, but I can't even find complete documentation for the HasuraCRUD schema. All I see is some descriptions + examples on the website. No formal documentation that would cover edge cases
